core/src/main/java/org/elasticsearch/rest/action/get/RestHeadSourceAction.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.rest.action.get;
+
+import org.elasticsearch.action.get.GetRequest;
+import org.elasticsearch.action.get.GetResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.rest.*;
+import org.elasticsearch.rest.action.support.RestResponseListener;
+
+import static org.elasticsearch.rest.RestRequest.Method.HEAD;
+import static org.elasticsearch.rest.RestStatus.NOT_FOUND;
+import static org.elasticsearch.rest.RestStatus.OK;
+
+/**
+ * Identical to {@link RestHeadAction} except this handler checks if the document and its _source exist.
+ */
+public class RestHeadSourceAction extends BaseRestHandler {
```

Comment:
> Are they run as part of gradle core:test?

No, the tests run as part of: `gradle :distribution:integ-test:integTest`. However I now realize that we don't have any rest integration test for HEAD actions... this doesn't seem to work at the moment too. So adding rest tests for head actions should be done another time.

> The GET actions were already split into 2 files: RestGetAction and RestGetSourceAction

True, these have different files, but the there also more difference between the two. The get request supports `fields` option while get source does not.

> Also request.path().endsWith("_source") is false for HEAD test-index/test-type/1/_source/ (and also for HEAD test-index/test-type/1/_source// which may be a bug?) 

Agreed, that does lead to bugs.

> I also had another idea to do something clever in the controller.registerHandler call to distinguish the HEAD doc request vs HEAD doc _source request. Maybe something like this?

I like that approach. However the `RestHeadAction.java` does need to extend from `RestHandler` otherwise it can't registered in `NetworkModule.java`. Maybe we should do something like [this](https://gist.github.com/martijnvg/6a28e57e6b39be2cdf4044a3dbc85669)? We would then also need to modify the  `NetworkModule.java` file to register the concrete subclasses instead.